### PR TITLE
Add TTY MultiSpinner

### PR DESCRIPTION
**Pull request by austb**
The only non-identical feature when compared to a single spinner is that clear doesn't remove the line entirely because the spinner has already printed a newline character. A newline character has to be printed (instead of cursor.down), otherwise when using a multispinner at the bottom of your terminal it will overwrite previous lines instead of creating new lines.

### CODE REVIEW

File: lib/tty/multi_spinner.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+      new_spinner
+    end
+
+    def lines_between_end_and_index(index)
+      count = 0
+      @spinners.each_with_index do |spinner, i|
+        next if i < index
+
+        count += 1 if spinner.spinning? || spinner.success? || spinner.error? || spinner.done?
```
Comment (piotrmurach) [inline lib/tty/multi_spinner.rb]: Why do we need to check for all these conditions? Is it to count all the spinners that are actually used/have been invoked by `auto_spin`, `run` etc..?

Comment (austb) [inline lib/tty/multi_spinner.rb]: Correct, spinning and done are straightforward, and when the spinner is entering a "success" or "error" state it is not yet "done" so those two are needed so the success/error message is printed on the correct line.

File: lib/tty/spinner.rb [hunk 2]
```diff
@@ -309,19 +325,45 @@ def update(tokens)
...
+      if @multispinner
+        LOCK.synchronize do
+          lines_up = @multispinner.lines_between_end_and_index(@index)
+
+          if @first_run
+            yield if block_given?
+            output.print "\n"
+            @first_run = false
+          else
+            output.print TTY::Cursor.save
```
Comment (piotrmurach) [inline lib/tty/spinner.rb]: I really like the idea of saving cursor and repainting what changed - nice one!

**Comment by piotrmurach** [comment]
I'm really impressed with this! Thanks for working on it. ❤️ 

I'm currently thinking about the use case and the api exposed by `TTY::MultiSpinner`. I'm thinking that generally people who decide to use multiple spinners will want to run them at once:
```ruby
spinners = TTY::MultiSpinner.new
# Tempted with TTY::Spinner::Multi.new
# as the TTY::Spinner is the main api entry

# add spinners and jobs
spinner1 = spinners.register("[:spinner] one") { puts "Performing work 1"; sleep(5); }
spinner2 = spinners.register("[:spinner] two") { puts "Performing work 2"; sleep(5); }
spinner3 = spinners.register("[:spinner] three") { puts "Performing work 3"; sleep(5); }

# Spinners would stop automatically once block finished but 
# you could cancel each spinner manually whenever by calling #stop

spinners.start # once setup all spinners, set them off
```

By using this slightly modified api we could remove all the spinner checks inside `lines_between_end_and_index`. What do you think?

**Comment by austb** [comment]
So for use case, we are using this at Puppet with [PDK](https://github.com/puppetlabs/pdk) and specifically [this PR](https://github.com/puppetlabs/pdk/pull/144). `pdk validate` validates three aspects of the [module](https://docs.puppet.com/puppet/5.0/modules_fundamentals.html) the Ruby code, the Puppet code, and the `metadata.json` file. We want to run these three checks in parallel, but for Puppet code and metadata.json validation it first validates the syntax (using a spinner) and if that is successful it then validates the style (adds another spinner). This is why I wanted the ability to add/start spinners dynamically.

**Comment by austb** [comment]
Also, I'm a big fan of the method name `register`, making that change now.

**Comment by piotrmurach** [comment]
Thanks for providing compelling use case :smile: I'm sold. I've quickly run through the Puppet PDK example PR and there is few things I think we could easily add:
* WindowsSpinner - I'm quite keen to support windows
* Auto switch unicode chars to ansii codes on windows
* Add more error messages for #on call and anywhere it makes sense

**Comment by austb** [comment]
We are actually in the process of [removing windows spinner](https://github.com/puppetlabs/pdk/pull/198). The main issue was that we weren't shipping aniscode with our MSI so it wasn't reliably available (we vendor all deps of the PDK for an MSI, ansicode should work on gem install though). Unicode support is a near impossibility on Windows, you can see some defaults we chose in that PR. Any thoughts on using those defaults?

**Comment by piotrmurach** [comment]
I've went through quite few pains to make `tty-prompt` work well on windows power shell. This means I've also have a bit of logic to [switch chars](https://github.com/piotrmurach/tty-prompt/blob/master/lib/tty/prompt/symbols.rb) on Windows such as ticks and crosses. I will probably extract that out to dependency as this is useful outside of `tty-prompt` lib.

**Comment by austb** [comment]
I currently left out .on(:done) .on(:success), etc because we don't have any use case for those once we have multiple spinners. We were using them to print out some more detailed information because we only had one spinner. Also I think there are some difficulties about when the Multi spinner is "done" because it can have spinners added any time.

Would you prefer to namespace `TTY::Spinner::Multi` to `TTY::MultiSpinner`?

**Comment by piotrmurach** [comment]
Do you think it would make sense to have `stop` call on multi spinner to stop all currently associated spinners? In this case, the done state could be easily determined.

Yes, let's change to `TTY::Spinner::Multi`, cheers!

How about changing `lines_between_end_and_index` to `count_lines_offset` and add some comments to explain the method intention? 

Anyways, thanks for this PR, it's midnight here so I will review and merge tomorrow.

**Comment by austb** [comment]
Yes, I would say it does (ie. we have to join our threads eventually to determine an exit code, so we can tell when we are done). I'll make that change, as well as the renaming and add some docs. Goodnight! Thank you very much for such a quick response.

**Comment by piotrmurach** [comment]
Thanks for your PR! I had a chance to review again and started doing some small refactoring of specs and multi spinner syntax. However, before I've started I've added [this example](https://github.com/piotrmurach/tty-spinner/blob/master/examples/multi.rb) which doesn't work at all in my terminal on mac os. The cursor save doesn't seem to remember any position so I get tons of new lines and spinners growing up the terminal window.

**Comment by austb** [comment]
Interesting, I am not seeing that problem here (also using Terminal.app on mac os).

**Comment by austb** [comment]
I've pushed my initial thoughts up in a commit [here](https://github.com/austb/tty-spinner/commit/43654cbbcc2f47acf2150d4101584bd6654241a2). Let me know if you see any other places we should protect.

**Comment by piotrmurach** [comment]
I've merged your fixes but I can still see the output traveling up my terminal overriding all prompts. See bellow:
screen shot 2017-07-28 at 22 44 09
As far as locks go, I think we need to have two locks, one to ensure that when updating output only one spinner is able to write to stdout and all others need to wait. Hence multispinner would need mutex that is exclusive that is also used by the current spinner when writing out to stdout. The second mutex would be used to protect spinner integrity. 

Having said all that, I just checked the example with just one spinner and it continues riding up the terminal screen. I think this is something to do with the newline and cursor position just not being right for some reason. Sorry, it's been long day I will have a look tomorrow.

**Comment by austb** [comment]
Ah I just realized why I didn't see the same thing, I was using tmux inside terminal.

**Pull request merged**